src/main/java/org/testng/internal/TestInvoker.java

```diff
@@ -259,15 +259,21 @@ private String checkDependencies(ITestNGMethod testMethod, ITestNGMethod[] allTe
           + "\"";
     }
 
+    if (!(context instanceof IConfigurationGroupMethodFetcher)) {
+      throw new TestNGException("context is not an instance of "
+          + IConfigurationGroupMethodFetcher.class.getName());
+    }
+
     // If this method depends on groups, collect all the methods that
     // belong to these groups and make sure they have been run successfully
     String[] groups = testMethod.getGroupsDependedUpon();
     if (null != groups && groups.length > 0) {
+    ITestNGMethod[] allTestMethods = ((IConfigurationGroupMethodFetcher) context).get().getAllTestMethods();
```

Comment:
Refactored this and got rid of `IConfigurationGroupMethodFetcher` completely. We now just call `context.getAllTestMethods()` which internally would figure out how to honour the invocation.
While a `TestRunner` is still active (i.e., its still executing something), we always return back what we have already cached via a configuration object and outside of its execution, we always compute the test methods. 
The code path here is within the TestRunner being active, so it will now just retrieve values from the configuration object and so this implementation is no longer relevant or needed.